### PR TITLE
PPF-229-settings tab

### PR DESCRIPTION
**Pull request by vhande**
### Added
-Alert component

### Fixed
-Settings Tab

success
error

---
Ticket: https://jira.uitdatabank.be/browse/PPF-229

### CODE REVIEW

File: app/Domain/Integrations/Models/IntegrationUrlModel.php [hunk 1]
```diff
@@ -29,10 +30,17 @@ final class IntegrationUrlModel extends UuidModel
     protected static function booted(): void
     {
         self::created(
-            static fn (IntegrationUrlModel $integrationUrlModel) => IntegrationUrlCreated::dispatch(
-                Uuid::fromString($integrationUrlModel->id),
-                Uuid::fromString($integrationUrlModel->integration_id)
-            )
+            static function (IntegrationUrlModel $integrationUrlModel) {
+                // @phpstan-ignore-next-line
+                $id = $integrationUrlModel->id instanceof LazyUuidFromString ? $integrationUrlModel->id : Uuid::fromString($integrationUrlModel->id);
```
Comment (LucWollants) [inline app/Domain/Integrations/Models/IntegrationUrlModel.php:35]: What is causing this `id` to have type `LazyUuidFromString`?

File: app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php [hunk 2]
```diff
@@ -20,9 +21,33 @@ public function rules(): array
         ];
 
         return [
-            'loginUrl' => $urlValidation,
+            'integrationName' => ['required', 'string', 'max:255'],
+            'description' => ['required', 'string', 'max:255'],
+            'loginUrls.*' => Rule::forEach(fn () => $urlValidation),
             'callbackUrls.*' => Rule::forEach(fn () => $urlValidation),
             'logoutUrls.*' => Rule::forEach(fn () => $urlValidation),
+            'newIntegrationUrls.*' => Rule::forEach(fn () => [
```
Comment (LucWollants) [inline app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php:29]: What kind of URLs are `newIntegrationUrls`?

File: app/Domain/Integrations/Models/IntegrationUrlModel.php (hunk 1 above)
Comment (simon-debruijn) [reply app/Domain/Integrations/Models/IntegrationUrlModel.php:35]: We investigated this, but it looks like an implementation from the uuid lib detail rather than our code

File: app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php (hunk 2 above)
Comment (simon-debruijn) [reply app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php:29]: Those are urls that have been added in the frontend in the current usage of the website. 
The state for already existing urls that are being edited is in separate fields from the newly added urls.

File: app/Domain/Integrations/Models/IntegrationUrlModel.php (hunk 1 above)
Comment (LucWollants) [reply app/Domain/Integrations/Models/IntegrationUrlModel.php:35]: Could it be that the new integration URLs sometimes have an empty ID?

File: app/Domain/Integrations/Controllers/IntegrationController.php [hunk 3]
```diff
@@ -140,41 +133,61 @@ public function destroyUrl(Request $request, string $id, string $urlId): Redirec
...
-            [
-                'id' => $id,
-            ]
-        );
+            $this->integrationRepository->update($updatedIntegration);
+        }
     }
 
-    public function updateUrls(UpdateIntegrationUrlsRequest $request, string $id): RedirectResponse
+    public function update(UpdateIntegrationUrlsRequest $request, string $id): RedirectResponse
```
Comment (LucWollants) [inline app/Domain/Integrations/Controllers/IntegrationController.php:147]: Was this name change intentional? Because the request has still type `UpdateIntegrationUrlsRequest` but the method name makes me think the full integration gets updated?

File: app/Domain/Integrations/Controllers/IntegrationController.php [hunk 4]
```diff
@@ -140,41 +133,61 @@ public function destroyUrl(Request $request, string $id, string $urlId): Redirec
...
+        $integrationUpdateRequest = new UpdateIntegrationRequest($integrationRequest);
+
+        $this->updateIntegration($integrationUpdateRequest, $id);
+
         $ids = array_map(
             fn ($url) => Uuid::fromString($url['id']),
             array_filter(
                 [
-                    $request->input('loginUrl'),
+                    ...($request->input('loginUrls') ?? []),
```
Comment (LucWollants) [inline app/Domain/Integrations/Controllers/IntegrationController.php:159]: I think in Auth0 there is only a single login URL

Comment (simon-debruijn) [reply app/Domain/Integrations/Controllers/IntegrationController.php:160]: Yes indeed loginUrl should be singular

File: app/Domain/Integrations/Controllers/IntegrationController.php [hunk 5]
```diff
@@ -142,41 +134,61 @@ public function destroyUrl(Request $request, string $id, string $urlId): Redirec
...
 
-        $this->integrationUrlRepository->updateUrls($updatedUrls);
+            $this->integrationUrlRepository->updateUrls($updatedUrls);
+        }
+
+        $newUrls =
+            array_filter(
+                [
+                    ...($request->input('newIntegrationUrls') ?? []),
+                ],
```
Comment (simon-debruijn) [inline app/Domain/Integrations/Controllers/IntegrationController.php:180]:
```suggestion
                ($request->input('newIntegrationUrls') ?? []),
```

No need to create a new array

File: app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php (hunk 2 above)
Comment (simon-debruijn) [reply app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php:29]: We better look into using upsert (update or insert depending on if it exists) logic

File: app/Domain/Integrations/Mappers/UpdateIntegrationUrlsMapper.php [hunk 6]
```diff
@@ -21,7 +21,7 @@ public static function map(UpdateIntegrationUrlsRequest $request, array $current
 
         $changedUrlsCollection = Collection::make(
             [
-                $request->input('loginUrl'),
+                ...($request->input('loginUrls') ?? []),
```
Comment (simon-debruijn) [inline app/Domain/Integrations/Mappers/UpdateIntegrationUrlsMapper.php:24]: Same remark about loginUrl being singular

File: resources/ts/Components/FormElement.tsx [hunk 7]
```diff
@@ -80,14 +81,18 @@ export const FormElement = ({
   component,
   error,
   info,
+  elementId,
```
Comment (simon-debruijn) [inline resources/ts/Components/FormElement.tsx:84]: Why is this needed?

File: resources/ts/Components/Integrations/Detail/IntegrationSettings.tsx [hunk 8]
```diff
@@ -90,6 +110,14 @@ export const IntegrationSettings = ({ integration, id, urls }: Props) => {
     setData("newIntegrationUrls", updated);
   };
 
+  const handleDeleteNewUrl = (fields?: string[], id?: string) => {
+    const updatedUrls = data.newIntegrationUrls
+      .filter((url) => fields?.includes(url.id))
+      .filter((url) => url.id !== id);
```
Comment (simon-debruijn) [inline resources/ts/Components/Integrations/Detail/IntegrationSettings.tsx:116]:
```suggestion
      .filter((url) => fields?.includes(url.id) && url.id !== id);
```

**Review by simon-debruijn — CHANGES_REQUESTED** [review]
I'll take a look how we can further improve simplify it when it is a upw week

File: resources/ts/Components/Integrations/Detail/UrlList.tsx [hunk 9]
```diff
@@ -86,71 +120,108 @@ export const UrlList = ({
...
-                </div>
-              )}
+                    )
+                )}
+              </div>
               {type !== IntegrationUrlType.Login && (
                 <ButtonSecondary
                   onClick={() => {
-                    option.changeVisibility(true);
+                    option.addNewUrlField(`${option.env + randomNumber}`);
```
Comment (brampauwelyn) [inline resources/ts/Components/Integrations/Detail/UrlList.tsx:224]: What is the reason of the `randomNumber`?

File: resources/ts/Components/Integrations/Detail/UrlList.tsx [hunk 10]
```diff
@@ -159,40 +230,107 @@ export const UrlList = ({
...
-                        type,
-                      })
-                    }
-                  />
-                </div>
-              }
-            />
+            <div key={option.env} className="flex flex-col gap-2">
+              <FormElement
+                label={`${t(`details.integration_settings.${option.env}`)}`}
```
Comment (brampauwelyn) [inline resources/ts/Components/Integrations/Detail/UrlList.tsx:235]: You can simplify this a bit:
```suggestion
                label={t(`details.integration_settings.${option.env}`)}
```

File: resources/ts/Components/Integrations/Detail/UrlList.tsx [hunk 11]
```diff
@@ -86,71 +120,108 @@ export const UrlList = ({
...
-                    `details.integration_settings.${url.environment}`
-                  )}`}
+                  elementId={url.id}
+                  className={`${type + option.env}`}
+                  error={errors[`${type}Urls.${`${url.id}`}.url`]}
+                  label={
+                    index === 0
+                      ? `${t(
+                          `details.integration_settings.${url.environment}`
+                        )}`
```
Comment (brampauwelyn) [inline resources/ts/Components/Integrations/Detail/UrlList.tsx:133]:
```suggestion
                      ? t(
                          `details.integration_settings.${url.environment}`
                        )
```

File: resources/ts/Components/Alert.tsx [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+        "flex gap-3 justify-between items-start rounded-md p-4",
+        variantToAlertStyle[variant],
+        className
+      )}
+    >
+      <div className={"flex gap-3 items-center"}>
+        <FontAwesomeIcon
+          icon={variantToIcon[variant]}
+          className={variantToIconColor[variant]}
+          size={"xl"}
```
Comment (brampauwelyn) [inline resources/ts/Components/Alert.tsx:71]:
```suggestion
          size="xl"
```

File: resources/ts/Components/Integrations/Detail/UrlList.tsx (hunk 9 above)
Comment (simon-debruijn) [reply resources/ts/Components/Integrations/Detail/UrlList.tsx:224]: The PR isn't ready for review yet, my bad.

This is remaining code I still have to refactor

File: resources/ts/Components/Integrations/Detail/UrlList.tsx (hunk 10 above)
Comment (simon-debruijn) [reply resources/ts/Components/Integrations/Detail/UrlList.tsx]: I no longer find this code

File: resources/ts/Components/Integrations/Detail/UrlList.tsx (hunk 11 above)
Comment (simon-debruijn) [reply resources/ts/Components/Integrations/Detail/UrlList.tsx]: I no longer find this code

**Review by simon-debruijn — APPROVED** [review]

File: app/Domain/Integrations/Repositories/EloquentIntegrationRepository.php [hunk 13]
```diff
@@ -45,14 +45,14 @@ public function save(Integration $integration): void
...
     {
-        IntegrationModel::query()
-            ->findOrFail($integration->id->toString())
-            ->update([
-                'type' => $integration->type,
-                'name' => $integration->name,
-                'description' => $integration->description,
-                'subscription_id' => $integration->subscriptionId->toString(),
-                'status' => $integration->status,
+        $integrationModel = IntegrationModel::query()->findOrFail($integration->id->toString());
```
Comment (LucWollants) [inline app/Domain/Integrations/Repositories/EloquentIntegrationRepository.php:48]: Is there a special need for the local variable `$integrationModel`?

File: app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php [hunk 14]
```diff
@@ -14,15 +18,36 @@ final class UpdateIntegrationUrlsRequest extends FormRequest
...
+            'urls.*' => Rule::forEach(fn (array $value, string $attribute) => [
+                'id' => ['string'],
+                'environment' => ["required_without:$attribute.id", new Enum(Environment::class)],
+                'type' => ["required_without:$attribute.id", new Enum(IntegrationUrlType::class)],
+                'url' => ['required', 'url:https', 'max:255'],
+            ]),
         ];
+    }
 
+    public function after(): array
```
Comment (LucWollants) [inline app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php:31]: This looks rather complex, by just looking at the code I don't know what it does 😅

Comment (simon-debruijn) [reply app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php:31]: The after method is called after the main validation is done.

The error keys that will be generated are in the following structure:
`urls.0.url`

We can't map these keys to the frontend fields. Because in the frontend we group and filter the urls based on the type and environment.

What this code does, is create new error keys with the following structure:
`type.environment.url`
We also have to use `getData` to use the correct type, environment and url per url object

And after adding out custom error keys, we forget (delete) the old error keys

Comment (simon-debruijn) [reply app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php:31]: In the laravel docs it says the after method should be replaced, by something else:
https://laravel.com/docs/10.x/upgrade#form-request-after-method

https://laravel.com/docs/10.x/validation#performing-additional-validation

But to me its unclear what the alternative is.

File: app/Domain/Integrations/Repositories/EloquentIntegrationRepository.php (hunk 13 above)
Comment (simon-debruijn) [reply app/Domain/Integrations/Repositories/EloquentIntegrationRepository.php:48]: I don't think so

Comment (simon-debruijn) [reply app/Domain/Integrations/Repositories/EloquentIntegrationRepository.php:48]: I see other instances in the code where it also defined first, to then call a method on it.
e.g.
```php
$integrationModel = IntegrationModel::query()->findOrFail($id->toString());
$integrationModel->activateWithOrganization($organizationId);
```

Comment (LucWollants) [reply app/Domain/Integrations/Repositories/EloquentIntegrationRepository.php:48]: But that line is to convert it a `IntegrationModel` and make PHPStan happy I think.

File: app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php (hunk 14 above)
Comment (LucWollants) [reply app/Domain/Integrations/FormRequests/UpdateIntegrationUrlsRequest.php]: Thanks for the extra comments. I think the first bigger comment was enough, the rest can be deduced from reading the code.

**Review by LucWollants — APPROVED** [review]
